angr/engines/pcode/lifter.py

```diff
@@ -0,0 +1,1506 @@
+        "data_refs",
+        "behaviors"
+    )
+
+    # The following constants shall match the defs in pyvex.h
+    MAX_EXITS = 400
+    MAX_DATA_REFS = 2000
+
+    def __init__(
+        self,
+        data,
+        mem_addr,
+        arch,
+        max_inst=None,
+        max_bytes=None,
+        bytes_offset=0,
+        traceflags=0,
+        opt_level=1,
+        num_inst=None,
+        num_bytes=None,
+        strict_block_end=False,
+        skip_stmts=False,
+        collect_data_refs=False,
+    ):
+        """
+        :param data:                The bytes to lift. Can be either a string of bytes or a cffi buffer object.
+                                    You may also pass None to initialize an empty IRSB.
+        :type data:                 str or bytes or cffi.FFI.CData or None
+        :param int mem_addr:        The address to lift the data at.
+        :param arch:                The architecture to lift the data as.
+        :type arch:                 :class:`archinfo.Arch`
+        :param max_inst:            The maximum number of instructions to lift. (See note below)
+        :param max_bytes:           The maximum number of bytes to use.
+        :param num_inst:            Replaces max_inst if max_inst is None. If set to None as well, no instruction limit is used.
+        :param num_bytes:           Replaces max_bytes if max_bytes is None. If set to None as well, no  byte limit is used.
+        :param bytes_offset:        The offset into `data` to start lifting at. Note that for ARM THUMB mode, both
+                                    `mem_addr` and `bytes_offset` must be odd (typically `bytes_offset` is set to 1).
+        :param traceflags:          Unused by P-Code lifter
+        :param opt_level:           Unused by P-Code lifter
+        :param strict_block_end:    Unused by P-Code lifter
+
+        .. note:: Explicitly specifying the number of instructions to lift (`max_inst`) may not always work
+                  exactly as expected. For example, on MIPS, it is meaningless to lift a branch or jump
+                  instruction without its delay slot. VEX attempts to Do The Right Thing by possibly decoding
+                  fewer instructions than requested. Specifically, this means that lifting a branch or jump
+                  on MIPS as a single instruction (`max_inst=1`) will result in an empty IRSB, and subsequent
+                  attempts to run this block will raise `SimIRSBError('Empty IRSB passed to SimIRSB.')`.
+
+        .. note:: If no instruction and byte limit is used, the lifter will continue lifting the block until the block
+                  ends properly or until it runs out of data to lift.
+        """
+        if max_inst is None:
+            max_inst = num_inst
+        if max_bytes is None:
+            max_bytes = num_bytes
+
+        self.addr = mem_addr
+        self.arch = arch
+
+        self._statements = []
+        self.next = None  # type: Optional[IRExpr]
```

Comment:
Let's use inline type annotations instead.

Reply:
I've just learned about this [syntax of variable annotations](https://www.python.org/dev/peps/pep-0526/) after reading your comment, thanks. I can do that.

Reply:
Done